code/iaas/api-logic/src/main/java/io/cattle/platform/iaas/api/auth/github/ProjectResourceManager.java

```diff
@@ -0,0 +1,140 @@
+    private static final String EXTERNAL_ID = "externalId";
+    private static final String EXTERNAL_ID_TYPE = "externalIdType";
+
+    JsonMapper jsonMapper;
+    GithubUtils githubUtils;
+    GithubClient githubClient;
+    AuthDao authDao;
+
+    @Override
+    public String[] getTypes() {
+        return new String[] { "project" };
+    }
+
+    @Override
+    public Class<?>[] getTypeClasses() {
+        return new Class[] {};
+    }
+
+    @Override
+    public Object listInternal(SchemaFactory schemaFactory, String type, Map<Object, Object> criteria, ListOptions options) {
+        HttpServletRequest request = ApiContext.getContext().getApiRequest().getServletContext().getRequest();
+        String token = request.getHeader(AUTH);
+        if (StringUtils.isEmpty(token)) {
+            return null;
+        }
+        Pair<String, Pair<List<String>, List<String>>> ids = githubUtils.validateAndFetchAccesibleIdsFromToken(token);
+        String userId = ids.getLeft();
+        Pair<List<String>, List<String>> accesibleids = ids.getRight();
+        List<String> teamIds = accesibleids.getLeft();
+        List<String> orgIds = accesibleids.getRight();
+        Account userAccount = authDao.getAccountByExternalId(userId, GITHUB_EXTERNAL_TYPE);
+        if (userAccount == null) {
+            return null;
+        }
+        String obfuscatedId = (String) ApiContext.getContext().getIdFormatter().formatId("account", Long.toString(userAccount.getId()));
+        return authDao.getAccessibleProjects(obfuscatedId, orgIds, teamIds);
+    }
+
+    @Override
+    protected Object createInternal(String type, ApiRequest request) {
+        if (!AccountConstants.TYPE_PROJECT.equals(type)) {
+            return null;
+        }
+
+        return createProject(request);
+    }
+
+    private Account createProject(ApiRequest apiRequest) {
+        HttpServletRequest request = apiRequest.getServletContext().getRequest();
+        String token = request.getHeader(AUTH);
+        if (StringUtils.isEmpty(token)) {
+            throw new ClientVisibleException(ResponseCodes.METHOD_NOT_ALLOWED);
+        }
+        Map<String, Object> account = CollectionUtils.toMap(apiRequest.getRequestObject());
+        String externalId = getExternalId((String) account.get(EXTERNAL_ID), (String) account.get(EXTERNAL_ID_TYPE), token);
+        if (StringUtils.isEmpty(externalId)) {
+            return null;
+        }
+        return authDao.createAccount((String) account.get("name"), "project", externalId, (String) account.get(EXTERNAL_ID_TYPE));
+    }
+
+    private String getExternalId(String externalId, String externalIdType, String jwt) {
+        String token = githubUtils.validateAndFetchGithubToken(jwt);
+        if (StringUtils.equals(externalIdType, TEAM_SCOPE)) {
+            if (StringUtils.isEmpty(externalId)) {
+                throw new ClientVisibleException(ResponseCodes.BAD_REQUEST, "MissingRequired", "externalId for TEAM scope should not be null", null);
```

Comment:
Already a constant for "MissingRequired" io.github.ibuildthecloud.gdapi.validation.ValidationErrorCodes.MISSING_REQUIRED